### PR TITLE
Refactor Hamming weight phasing to use QVR

### LINKED ISSUE #1218: Unify `HammingWeightPhasing` bloqs
[hamming_weight_phasing.py](https://github.com/quantumlib/Qualtran/blob/2a78ca1d3641dcfce4c4813456dfdee1c6c00749/qualtran/bloqs/rotations/hamming_weight_phasing.py)

The two variants [`HammingWeightPhasing`](https://github.com/quantumlib/Qualtran/blob/2a78ca1d3641dcfce4c4813456dfdee1c6c00749/qualtran/bloqs/rotations/hamming_weight_phasing.py#L33) and [`HammingWeightPhasingViaPhaseGradient`](https://github.com/quantumlib/Qualtran/blob/2a78ca1d3641dcfce4c4813456dfdee1c6c00749/qualtran/bloqs/rotations/hamming_weight_phasing.py#L123) are nearly identical, except the Z rotation part. This should be a single bloq with an [`QVRInterface`](https://github.com/quantumlib/Qualtran/blob/2a78ca1d3641dcfce4c4813456dfdee1c6c00749/qualtran/bloqs/rotations/quantum_variable_rotation.py#L96) bloq attribute to perform the rotation.

### PR DESCRIPTION
**Pull request by anurudhp**
fixes #1218

### CODE REVIEW

File: qualtran/bloqs/rotations/hamming_weight_phasing.py [hunk 1]
```diff
@@ -74,24 +78,77 @@ class HammingWeightPhasing(GateWithRegisters):
...
     @cached_property
     def signature(self) -> 'Signature':
-        return Signature.build_from_dtypes(x=QUInt(self.bitsize))
+        return Signature([Register("x", QUInt(self.bitsize)), *self.phase_oracle.extra_registers])
+
+    @classmethod
+    def via_phase_gradient(cls, bitsize: int, exponent: float = 1, eps: SymbolicFloat = 1e-10):
+        r"""Factory method for `HammingWeightPhasing` using a phase gradient state.
+
+        In this variant of Hamming Weight Phasing, instead of directly synthesizing $O(\log_2 (\texttt{bitsize}))$
```
Comment (mpharrigan) [inline qualtran/bloqs/rotations/hamming_weight_phasing.py:99]: I know you copied this, but can you hard-wrap the text to <100 columns. vim is `gq` ; pycharm is "edit -> fill paragraph"

File: qualtran/bloqs/rotations/hamming_weight_phasing.py [hunk 2]
```diff
@@ -74,24 +78,77 @@ class HammingWeightPhasing(GateWithRegisters):
...
+            exponent: The exponent of `Z ** exponent` to be applied to each qubit in the input register.
+            eps: Accuracy of synthesizing the Z rotations.
+
+        Registers:
+            x : Input THRU register of size `bitsize`, to apply `Z**exponent` to.
+            phase_grad : Phase gradient THRU register of size `O(log2(1/eps))`, to be used to
+                apply the phasing via addition.
+
+        References:
+            1. [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization]
```
Comment (mpharrigan) [inline qualtran/bloqs/rotations/hamming_weight_phasing.py:127]: I know you copied this, but remove `1. ` from the list of references. We don't number them.

File: qualtran/bloqs/rotations/hamming_weight_phasing.py [hunk 3]
```diff
@@ -74,24 +78,77 @@ class HammingWeightPhasing(GateWithRegisters):
...
+            bitsize: Size of input register to apply `Z ** exponent` to.
+            exponent: The exponent of `Z ** exponent` to be applied to each qubit in the input register.
+            eps: Accuracy of synthesizing the Z rotations.
+
+        Registers:
+            x : Input THRU register of size `bitsize`, to apply `Z**exponent` to.
+            phase_grad : Phase gradient THRU register of size `O(log2(1/eps))`, to be used to
+                apply the phasing via addition.
+
+        References:
```
Comment (mpharrigan) [inline qualtran/bloqs/rotations/hamming_weight_phasing.py:126]: consider moving the reference to the class docstring

File: qualtran/bloqs/rotations/hamming_weight_phasing.py [hunk 4]
```diff
@@ -66,6 +69,7 @@ class HammingWeightPhasing(GateWithRegisters):
 
     Registers:
         x: A `THRU` register of `bitsize` qubits.
+        extra_registers: Any additional registers used by the `QvrInterface` phase oracle.
```
Comment (mpharrigan) [inline qualtran/bloqs/rotations/hamming_weight_phasing.py:72]: document the new `phase_oracle` arg and reference the new factory method in the class docstring somewhere

**Comment by mpharrigan** [comment]
@tanujkhattar original author of these bloqs: are you good with the merger?

**Comment by mpharrigan** [comment]
@tanujkhattar @anurudhp what's the status of this?

**Comment by anurudhp** [comment]
This has deviated too much from main, closing it for now. We can make a new refactor PR if needed

**Pull request closed without merging**